### ISSUE TITLE
Specify TSC Frequency As Kernel Boot Parameter

### ISSUE THREAD

**Issue by alexandruag** [firecracker-microvm/firecracker#477]
We should look up that boot parameter which allows us to inform the kernel about TSC frequency so it can skip a bunch of initialization, and maybe we lose ~30ms of boot latency.

**Comment by cneira** [comment]
@alexandruag , @andreeaflorescu 
The intent of this ticket is to add clocksource as a default boot parameter ? 
```rust
const DEFAULT_KERNEL_CMDLINE: &str = "reboot=k panic=1 pci=off nomodules 8250.nr_uarts=0 clocksource=tsc";
```
without clocksource=tsc
```bash
   0.000000] Linux version 4.14.55-84.37.amzn2.x86_64 (mockbuild@ip-10-0-1-79) (gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC)) #1 SMP Wed Jul 25 18:47:15 UTC 2018
[    0.000000] Command line: console=ttyS0 reboot=k panic=1 pci=off  root=/dev/vda virtio_mmio.device=4K@0xd0000000:5
[    0.000000] x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x008: 'MPX bounds registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x010: 'MPX CSR'
[    0.000000] x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
[    0.000000] x86/fpu: xstate_offset[3]:  832, xstate_sizes[3]:   64
[    0.000000] x86/fpu: xstate_offset[4]:  896, xstate_sizes[4]:   64
[    0.000000] x86/fpu: Enabled xstate features 0x1f, context size is 960 bytes, using 'compacted' format.
[    0.000000] e820: BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x0000000007ffffff] usable
[    0.000000] NX (Execute Disable) protection: active
[    0.000000] DMI not present or invalid.
[    0.000000] Hypervisor detected: KVM
[    0.000000] tsc: Fast TSC calibration failed
[    0.000000] tsc: Unable to calibrate against PIT
[    0.000000] tsc: No reference (HPET/PMTIMER) available
[    0.000000] e820: last_pfn = 0x8000 max_arch_pfn = 0x400000000
[    0.000000] MTRR: Disabled
```
with clocksource=tsc
```bash
[    0.000000] Linux version 4.14.55-84.37.amzn2.x86_64 (mockbuild@ip-10-0-1-79) (gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC)) #1 SMP Wed Jul 25 18:47:15 UTC 2018
[    0.000000] Command line: console=ttyS0 reboot=k panic=1 pci=off clocksource=tsc  root=/dev/vda virtio_mmio.device=4K@0xd0000000:5
[    0.000000] x86/fpu: Supporting XSAVE feature 0x001: 'x87 floating point registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x002: 'SSE registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x004: 'AVX registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x008: 'MPX bounds registers'
[    0.000000] x86/fpu: Supporting XSAVE feature 0x010: 'MPX CSR'
[    0.000000] x86/fpu: xstate_offset[2]:  576, xstate_sizes[2]:  256
[    0.000000] x86/fpu: xstate_offset[3]:  832, xstate_sizes[3]:   64
[    0.000000] x86/fpu: xstate_offset[4]:  896, xstate_sizes[4]:   64
[    0.000000] x86/fpu: Enabled xstate features 0x1f, context size is 960 bytes, using 'compacted' format.
[    0.000000] e820: BIOS-provided physical RAM map:
[    0.000000] BIOS-e820: [mem 0x0000000000000000-0x000000000009fbff] usable
[    0.000000] BIOS-e820: [mem 0x0000000000100000-0x0000000007ffffff] usable
[    0.000000] NX (Execute Disable) protection: active
[    0.000000] DMI not present or invalid.
[    0.000000] Hypervisor detected: KVM
[    0.000000] tsc: Fast TSC calibration using PIT
[    0.000000] e820: last_pfn = 0x8000 max_arch_pfn = 0x400000000
[    0.000000] MTRR: Disabled
```

**Comment by rbradford** [comment]
It is worth noting that with a newer kernel (e.g. 4.19) this delay doesn't happen; all that code has been heavily refactored. I was seeing the same problem with our "virt" machine platform in NEMU and changing the kernel version mitigated that. You can see the statistics and some details here in the work I collaborated with Stefano on: https://lists.gnu.org/archive/html/qemu-devel/2018-12/msg03096.html

**Comment by cneira** [comment]
Awesome. This could be closed then.

**Comment by serban300** [comment]
Listing the boot time for 10 runs with Linux Kernel 4.14.55-84.37:

```
Guest-boot-time = 275411 us 275 ms, 268687 CPU us 268 CPU ms
Guest-boot-time = 264406 us 264 ms, 259741 CPU us 259 CPU ms
Guest-boot-time = 256350 us 256 ms, 253783 CPU us 253 CPU ms
Guest-boot-time = 269144 us 269 ms, 261290 CPU us 261 CPU ms
Guest-boot-time = 270703 us 270 ms, 268118 CPU us 268 CPU ms
Guest-boot-time = 270550 us 270 ms, 265055 CPU us 265 CPU ms
Guest-boot-time = 260472 us 260 ms, 256911 CPU us 256 CPU ms
Guest-boot-time = 276310 us 276 ms, 267120 CPU us 267 CPU ms
Guest-boot-time = 271196 us 271 ms, 268849 CPU us 268 CPU ms
Guest-boot-time = 276590 us 276 ms, 266897 CPU us 266 CPU ms
```

**Comment by serban300** [comment]
It seems that the problem has been fixed in Kernel >= 4.19. Closing the issue.
